lib/Conversion/LTLToCore/LTLToCore.cpp

```diff
@@ -0,0 +1,237 @@
+    // Match for a single verif.assert user, and replace the structure.
+    // Folding will recursively apply this pattern if multiple instances are
+    // found.
+    auto users = op.getResult().getUsers();
+    if (users.empty())
+      return rewriter.notifyMatchFailure(op, "No users found");
+
+    for (auto *user : op.getResult().getUsers()) {
+      auto fusedLoc =
+          mlir::FusedLoc::get(getContext(), {op.getLoc(), user->getLoc()});
+
+      // Dispatch to the clock-specific patterns.
+      LogicalResult res =
+          llvm::TypeSwitch<Operation *, LogicalResult>(user)
+              .Case<verif::AssertOp>([&](auto assert) {
+                return rewriteAssertOp(fusedLoc, assert, adaptor, rewriter);
+              })
+              .Default([&](auto) {
+                return rewriter.notifyMatchFailure(
+                    op, "User of clock is not a verif.assert");
+              });
+      if (failed(res))
+        return res;
+    }
+
+    // Clock can be removed as all users should have been handled by now.
+    rewriter.eraseOp(op);
+    return success();
+  }
+
+  LogicalResult rewriteAssertOp(Location loc, verif::AssertOp assertOp,
+                                OpAdaptor &adaptor,
+                                ConversionPatternRewriter &rewriter) const {
+    // Generate the parenting sv.always posedge clock from the ltl
+    // clock, containing the generated sv.assert
+    rewriter.replaceOpWithNewOp<sv::AlwaysOp>(
+        assertOp, (sv::EventControl)adaptor.getEdge(), adaptor.getClock(), [&] {
+          // Generate the sv assertion using the input to the parenting
+          // clock
+          rewriter.create<sv::AssertOp>(
+              loc, adaptor.getInput(),
+              sv::DeferAssertAttr::get(getContext(),
+                                       sv::DeferAssert::Immediate),
+              assertOp.getLabelAttr());
+        });
+
+    return success();
+  }
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Lower LTL To Core pass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LowerLTLToCorePass : public LowerLTLToCoreBase<LowerLTLToCorePass> {
+  LowerLTLToCorePass() = default;
+  void runOnOperation() override;
+};
+} // namespace
+
+// Applies all of the conversion patterns needed for this lowering
+void circt::populateLTLToCoreConversionPatterns(
+    circt::LTLToCoreTypeConverter &converter,
+    mlir::RewritePatternSet &patterns) {
+  patterns.add<HasBeenResetOpConversion, LowerClockRelatedOpPatterns,
+               DisableOpConversion>(converter, patterns.getContext());
+}
+
+// Creates the type conversions and materializations needed for this pass to
+// work
+circt::LTLToCoreTypeConverter::LTLToCoreTypeConverter() {
+
+  // Convert the ltl property type to a built-in type
+  addConversion([](Type type) { return type; });
```

Comment:
It wouldn't verify things otherwise, but that might of also been due to something else. I will try to get rid of it and see what happens.

Reply:
Yes, it won't work otherwise

Reply:
I think this is not quite right. This is saying any Type is legal. I think what you want is `addConversion([](IntegerType type) { return type; })`, which says IntegerType is legal. You do need that, since you want to convert PropertyType to IntegerType, but then you want the conversion framework to understand that IntegerType is legal. This might be related to the other comment thread where you mentioned converting backwards didn't work.